cmd/proxy.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/cloudsqlconn"
+)
+
+type socketMount struct {
+	dialer cloudsqlconn.Dialer
+	inst   string
+
+	listener net.Listener
+}
+
+func newSocketMount(dialer cloudsqlconn.Dialer, inst string) *socketMount {
+	return &socketMount{
+		dialer: dialer,
+		inst:   inst,
+	}
+}
+
+func (s *socketMount) Listen(ctx context.Context, network string, host string) (net.Addr, error) {
```

Comment:
Why export any of these methods when the type and its constructor as unusable for public consumers?

Reply:
I think we've discussed this before, but this is a kvg-ism that helps me associate between "internal" methods on the struct and methods on the struct I want to be able to use externally. 

If it bothers you I can change, but it helps me keep it straight a bit more.

Reply:
We could also move this code into an `internal` directory and make it properly exported. It's very confusing to see exported methods that are in fact private.